File: src/main/java/raylras/zen/langserver/ZenLanguageServer.java
package raylras.zen.langserver;

import org.eclipse.lsp4j.*;
import org.eclipse.lsp4j.jsonrpc.messages.Either;
import org.eclipse.lsp4j.services.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import raylras.zen.code.CompilationUnit;
import raylras.zen.util.l10n.L10N;

import java.util.*;
import java.util.concurrent.CompletableFuture;

public class ZenLanguageServer implements LanguageServer, LanguageClientAware {

    private static final Logger logger = LoggerFactory.getLogger(ZenLanguageServer.class);

    private final ZenLanguageService service;

    public ZenLanguageServer() {
        this.service = new ZenLanguageService();
    }

    @Override
    public void setTrace(SetTraceParams params) {
    }

    @Override
    public CompletableFuture<InitializeResult> initialize(InitializeParams params) {
        service.initializeWorkspaces(params.getWorkspaceFolders());
        L10N.setLocale(params.getLocale());

        ServerCapabilities capabilities = new ServerCapabilities();
        capabilities.setTextDocumentSync(TextDocumentSyncKind.Full);
        capabilities.setCompletionProvider(new CompletionOptions(true, List.of(".", ":")));
        capabilities.setDocumentSymbolProvider(true);
        // capabilities.setWorkspaceSymbolProvider(true);
        // capabilities.setDocumentHighlightProvider(true);
        // SignatureHelpOptions signatureHelpOptions = new SignatureHelpOptions();
        // signatureHelpOptions.setTriggerCharacters(Arrays.asList("(", ","));
        // capabilities.setSignatureHelpProvider(signatureHelpOptions);
        // capabilities.setSemanticTokensProvider(new SemanticTokensWithRegistrationOptions(Semantics.SEMANTIC_TOKENS_LEGEND, true));
        capabilities.setReferencesProvider(true);
        // capabilities.setDeclarationProvider(true);
        capabilities.setDefinitionProvider(true);
        // capabilities.setTypeDefinitionProvider(true);
        capabilities.setHoverProvider(true);
        // capabilities.setRenameProvider(true);
        return CompletableFuture.completedFuture(new InitializeResult(capabilities));
    }

    @Override
    public void initialized(InitializedParams params) {
        startListeningFileChanges();
        logger.info("Language server initialized");
    }

    @Override
    public TextDocumentService getTextDocumentService() {
        return service;
    }

    @Override
    public WorkspaceService getWorkspaceService() {
        return service;
    }

    @Override
    public CompletableFuture<Object> shutdown() {
        return new CompletableFuture<>();
    }

    @Override
    public void exit() {
        System.exit(0);
    }

    @Override
    public void connect(LanguageClient client) {
        ZenLanguageService.setClient(client);
    }

    private void startListeningFileChanges() {
        List<FileSystemWatcher> watchers = new ArrayList<>(1);
        watchers.add(new FileSystemWatcher(Either.forLeft("**/*" + CompilationUnit.ZS_FILE_EXTENSION), WatchKind.Create + WatchKind.Change + WatchKind.Delete));
        Object options = new DidChangeWatchedFilesRegistrationOptions(watchers);
        Registration registration = new Registration(UUID.randomUUID().toString(), "workspace/didChangeWatchedFiles", options);
        ZenLanguageService.getClient().registerCapability(new RegistrationParams(Collections.singletonList(registration)));
    }

}

File: language-server/src/main/kotlin/raylras/intellizen/lsp/ZenLanguageServer.kt
package raylras.intellizen.lsp

import org.eclipse.lsp4j.DidChangeWatchedFilesRegistrationOptions
import org.eclipse.lsp4j.FileSystemWatcher
import org.eclipse.lsp4j.InitializeParams
import org.eclipse.lsp4j.InitializeResult
import org.eclipse.lsp4j.InitializedParams
import org.eclipse.lsp4j.Registration
import org.eclipse.lsp4j.RegistrationParams
import org.eclipse.lsp4j.ServerCapabilities
import org.eclipse.lsp4j.SetTraceParams
import org.eclipse.lsp4j.TextDocumentSyncKind
import org.eclipse.lsp4j.WatchKind
import org.eclipse.lsp4j.jsonrpc.messages.Either
import org.eclipse.lsp4j.services.LanguageServer
import org.slf4j.LoggerFactory
import raylras.intellizen.i18n.L10N
import java.util.UUID
import java.util.concurrent.CompletableFuture
import kotlin.apply
import kotlin.jvm.java
import kotlin.system.exitProcess

private val logger = LoggerFactory.getLogger(ZenLanguageServer::class.java)

class ZenLanguageServer(private val service: ZenLanguageService) : LanguageServer {
    override fun initialize(params: InitializeParams): CompletableFuture<InitializeResult> {
        L10N.setLocale(params.locale)
        val capabilities = ServerCapabilities().apply {
            setTextDocumentSync(TextDocumentSyncKind.Full)
//            setDocumentSymbolProvider(true)
//            setWorkspaceSymbolProvider(true)
//            setDocumentHighlightProvider(true)
//            setReferencesProvider(true)
//            setDeclarationProvider(true)
//            setDefinitionProvider(true)
//            setTypeDefinitionProvider(true)
//            setHoverProvider(true)
//            setRenameProvider(true)
//            completionProvider = CompletionOptions(true, listOf(".", "<"))
//            signatureHelpProvider = SignatureHelpOptions(listOf("(", ","))
//            semanticTokensProvider = SemanticTokensWithRegistrationOptions(SEMANTIC_TOKENS_LEGEND, false, true)
//            setInlayHintProvider(true)
        }
        return CompletableFuture.completedFuture(InitializeResult(capabilities))
    }

    override fun initialized(params: InitializedParams) {
        startListeningFileChanges()
        logger.info("Language server initialized")
    }

    override fun shutdown(): CompletableFuture<Any> {
        return CompletableFuture.supplyAsync {
            logger.info("Language server shutting down")
            StandardIOLauncher.shutdown()
        }
    }

    override fun exit() {
        logger.info("Language server exiting")
        exitProcess(0)
    }

    override fun getTextDocumentService() = service

    override fun getWorkspaceService() = service

    override fun setTrace(params: SetTraceParams) {
    }

    private fun startListeningFileChanges() {
        val watcher = listOf(
            FileSystemWatcher(
                Either.forLeft("**/*$.zs"),
                WatchKind.Create + WatchKind.Change + WatchKind.Delete
            )
        )
        val options = DidChangeWatchedFilesRegistrationOptions(watcher)
        val registration = Registration(UUID.randomUUID().toString(), "workspace/didChangeWatchedFiles", options)
        StandardIOLauncher.client?.registerCapability(RegistrationParams(listOf(registration)))
    }
}